consensus/istanbul/core/backlog.go

```diff
@@ -95,115 +108,178 @@ func (c *core) storeBacklog(msg *istanbul.Message, src istanbul.Validator) {
 		return
 	}
 
-	logger.Trace("Store future message")
-
-	c.backlogsMu.Lock()
-	defer c.backlogsMu.Unlock()
-
-	backlog := c.backlogs[src]
-	if backlog == nil {
-		backlog = prque.New(nil)
-	}
+	var v *istanbul.View
 	switch msg.Code {
 	case istanbul.MsgPreprepare:
 		var p *istanbul.Preprepare
 		err := msg.Decode(&p)
-		if err == nil {
-			backlog.Push(msg, toPriority(msg.Code, p.View))
+		if err != nil {
+			return
 		}
+		v = p.View
 	case istanbul.MsgPrepare:
 		fallthrough
 	case istanbul.MsgCommit:
 		var p *istanbul.Subject
 		err := msg.Decode(&p)
-		if err == nil {
-			backlog.Push(msg, toPriority(msg.Code, p.View))
+		if err != nil {
+			return
 		}
+		v = p.View
 	case istanbul.MsgRoundChange:
 		var p *istanbul.RoundChange
 		err := msg.Decode(&p)
-		if err == nil {
-			backlog.Push(msg, toPriority(msg.Code, p.View))
+		if err != nil {
+			return
+		}
+		v = p.View
+	}
+
+	logger.Trace("Store future message", "msg", msg)
+
+	c.backlogsMu.Lock()
+	defer c.backlogsMu.Unlock()
+
+	// Check and inc per-validator future message limit
+	if c.backlogCountByVal[msg.Address] > acceptMaxFutureMsgsFromOneValidator {
+		logger.Trace("Dropping: backlog exceeds per-src cap", "src", src)
+		return
+	}
+	c.backlogCountByVal[src.Address()]++
+	c.backlogTotal++
+
+	// Add message to per-seq list
+	backlogForSeq := c.backlogBySeq[v.Sequence.Uint64()]
+	if backlogForSeq == nil {
+		backlogForSeq = prque.New(nil)
+		c.backlogBySeq[v.Sequence.Uint64()] = backlogForSeq
+	}
+
+	backlogForSeq.Push(msg, toPriority(msg.Code, v))
+
+	// Keep backlog below total max size by pruning future-most sequence first
+	// (we always leave one sequence's entire messages and rely on per-validator limits)
+	if c.backlogTotal > acceptMaxFutureMessages {
+		backlogSeqs := c.getSortedBacklogSeqs()
+		for i := len(backlogSeqs) - 1; i > 0; i-- {
+			seq := backlogSeqs[i]
+			if seq <= c.currentView().Sequence.Uint64() ||
+				c.backlogTotal < (acceptMaxFutureMessages-acceptMaxFutureMessagesPruneBatch) {
+				break
+			}
+			c.drainBacklogForSeq(seq, nil)
 		}
 	}
-	c.backlogs[src] = backlog
+}
+
+// Return slice of sequences present in backlog sorted in ascending order
+// Call with backlogsMu held.
+func (c *core) getSortedBacklogSeqs() []uint64 {
+	backlogSeqs := make([]uint64, len(c.backlogBySeq))
+	i := 0
+	for k := range c.backlogBySeq {
+		backlogSeqs[i] = k
+		i++
+	}
+	sort.Slice(backlogSeqs, func(i, j int) bool {
+		return backlogSeqs[i] < backlogSeqs[j]
+	})
+	return backlogSeqs
+}
+
+// Drain a backlog for a given sequence, passing each to optional callback.
+// Call with backlogsMu held.
+func (c *core) drainBacklogForSeq(seq uint64, cb func(*istanbul.Message, istanbul.Validator)) {
+	backlogForSeq := c.backlogBySeq[seq]
+	if backlogForSeq == nil {
+		return
+	}
+
+	for !backlogForSeq.Empty() {
```

Comment:
check the size here, only iterate that many times, and afterwards only delete if it's empty.